### PR TITLE
Start ocs integration tests

### PR DESCRIPTION
**Pull request by DeepDiver1975**
refs https://github.com/owncloud/notifications/issues/45

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -44,6 +44,10 @@ script:
   - cd tests/
   - sh -c "if [ '$JSTESTS' != '1' -a '$CODECHECK' != '1' -a '$CODECHECK' != '2' ]; then phpunit --configuration phpunit.xml; fi"
 
+  # Integration tests
+  - cd tests/integration
```
Comment (nickvergessen) [inline .travis.yml]: we are already in `tests/` and afterwards we need to go back to it for the coverage command?

File: tests/integration/features/noticifations.feature [hunk 2]
```diff
@@ -0,0 +1,11 @@
+Feature: notification
```
Comment (nickvergessen) [inline tests/integration/features/noticifations.feature]: file name is :urgh:

**Comment by DeepDiver1975** [comment]
bäääh - phpunit is trying to load the folder integration ....

File: tests/integration/run.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
```
Comment (nickvergessen) [inline tests/integration/run.sh]: should `set -e` so it fails, I think?

**Comment by nickvergessen** [comment]
:+1: thanks

File: tests/integration/features/notifications.feature [hunk 4]
```diff
@@ -0,0 +1,21 @@
+Feature: notifications
+  Background:
+    Given using api version "1"
+
+  Scenario: Create test user
```
Comment (DeepDiver1975) [inline tests/integration/features/notifications.feature]: scenarios are to be self contained - "Creating a new notification" relies on "Create test user"

further more "Create test user" has nothing to do with notifications.

That a given user exists is a precondition to the scenario - this is why the original implementation did use

Given user "user01" exists

File: tests/integration/features/notifications.feature [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    When sending "POST" to "/cloud/users" with
+      | userid | test1 |
+      | password | 123456 |
+    Then the OCS status code should be "100"
+    And the HTTP status code should be "200"
+    And user "test1" exists
+
+  Scenario: Creating a new notification
+    Given As an "test1"
+    When sending "GET" to "/notifications/v1"
```
Comment (DeepDiver1975) [inline tests/integration/features/notifications.feature]: GET to create? :wink:

**Comment by nickvergessen** [comment]
Okay, so I changed some bits:
1. Fixed the path of the php server in run.sh
2. Load composer in before step so it does not pollute the output on travis
3. Copied the build/integration/ version to CoreContext and made [FeatureContext](https://github.com/owncloud/notifications/pull/46/files#diff-72f9ffe4450f86f8977b79d5fac91d13) inherit it
  * Changed class name to CoreContext
  * Commented out `addFilesToSkeleton` and `removeFilesFromSkeleton` because of the wrong paths

To have easy integration tests across apps, the following should happen:
1. Split FeatureContext into stuff that everyone needs and stuff that is core only (currently only the skeleton)
2. App developer copied CoreContext from build/integration/ into his folder before execution (e.g. on travis)

This way we don't duplicate it all the time and suffer from missing changes etc.

**Comment by nickvergessen** [comment]
If that's okay by you @DeepDiver1975 I'd like to go ahead and merge this.
Currently it tests the "no notifier" case successfully.
I will then continue working on integration tests for the other stuff on top of the new master, instead of ending up with a huge PR

File: appinfo/routes.php [hunk 6]
```diff
@@ -23,19 +23,19 @@
 
 \OCP\API::register(
 	'get',
-	'/notifications/v1',
+	'/apps/notifications/v1/notifications',
```
Comment (DeepDiver1975) [inline appinfo/routes.php]: apps/notification will be added by ocs already???

File: tests/integration/features/bootstrap/CoreContext.php [hunk 7]
```diff
@@ -0,0 +1,930 @@
...
+use Behat\Behat\Context\SnippetAcceptingContext;
+use GuzzleHttp\Client;
+use GuzzleHttp\Message\ResponseInterface;
+
+require __DIR__ . '/../../vendor/autoload.php';
+
+/**
+ * Features context.
+ */
+class CoreContext implements Context, SnippetAcceptingContext {
```
Comment (DeepDiver1975) [inline tests/integration/features/bootstrap/CoreContext.php]: oh - by simply copy-it-over I did mean to copy those methods which are needed :see_no_evil:

File: appinfo/routes.php (hunk 6 above)
Comment (nickvergessen) [inline appinfo/routes.php]: No https://github.com/owncloud/core/blob/master/apps/files_sharing/appinfo/routes.php#L77

File: tests/integration/features/bootstrap/CoreContext.php (hunk 7 above)
Comment (nickvergessen) [inline tests/integration/features/bootstrap/CoreContext.php]: The problem is it's like a rat tail... you end up copying more and more, and then it's also harder to track changes that have been done in core

**Pull request merged**